### ISSUE TITLE
Add typing support

### ISSUE THREAD

**Issue by palfrey** [confluentinc/confluent-kafka-python#1408]
Description
===========

Currently, if you run `mypy` with a project with this in, you get `error: Skipping analyzing "confluent_kafka.admin": module is installed, but missing library stubs or py.typed marker  [import]` or similar. https://mypy.readthedocs.io/en/stable/installed_packages.html has a guide on the ways to fix this.

How to reproduce
================

`mypy <some code that uses confluent_kafka>`

Checklist
=========
Please provide the following information:

 - [x] confluent-kafka-python and librdkafka version (`confluent_kafka.version()` and `confluent_kafka.libversion()`): 1.9.2
 - [ ] Apache Kafka broker version:
 - [ ] Client configuration: `{...}`
 - [ ] Operating system:
 - [ ] Provide client logs (with `'debug': '..'` as necessary)
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by mhowlett** [comment]
thanks for the suggestion @palfrey - yep, i would love to get more "polish the cover of the book" type things done.

prioritization coming down from product is all internal atm.

**Comment by LiamClarkeNZ** [comment]
Very much a +1 on this. 

@jliunyu previously stated in #1268 that this is low on the agenda for Confluent as it doesn't offer value to Python users on < 3.5, but the fact that `confluent-kafka-python` is backed by Confluent is why we used it instead of `kafka-python`, as we're signed up onto Confluent Cloud with its associated support, but it's not super-great when your IDE thinks that `kafka_message.value()` is missing a keyword argument of `payload`, as it introduces a lot of false positive warnings into code using `confluent-kafka-python`. 

I've worked around it somehow with mypy's stubgen and manual editing of .pyi files based on API documentation, but when the API documentation (for a recent example) doesn't describe the type parameters of the `on_assign` callback to `subscribe` then it gets slow and painful to get the type hints correct.

**Comment by palfrey** [comment]
https://github.com/confluentinc/confluent-kafka-python/pull/1493 is an initial shot at this

**Comment by Kyle-sandeman-mrdfood** [comment]
Hi, just to add a comment about `message.value()` in PyCharm, it seems that in the docstring `value (payload)` is interpreted as "function signature docs".
My suggestion would be `value (the payload)` (cannot verify whether this works right now) but it feels a bit hacky. My guess is that PyCharm looks for a valid variable name (or comma-separated ones) inside brackets.
The "proper" solution would be creating a `cimpl.pyi` file with the correct types, although this is a maintenance headache going forward.
Not sure about backwards compatibility for < 3.4 or 2.x but as palfrey says in the PR, this may not be required.

**Comment by olejorgenb** [comment]
From #1268 (jan. 2022)

> But since the type hints are supporting from python 3.5, it's not compatible with python version lower than 3.5, we still have a lot of customers using python version lower than 3.5, it's hard for us to import this feature for now. But we may consider this feature later.

Time to reconsider in 2024?

**Comment by LiamClarkeNZ** [comment]
> From #1268 (jan. 2022)
> 
> > But since the type hints are supporting from python 3.5, it's not compatible with python version lower than 3.5, we still have a lot of customers using python version lower than 3.5, it's hard for us to import this feature for now. But we may consider this feature later.
> 
> Time to reconsider in 2024?

Especially as type hints can be provided via Typeshed, they don't need to be in the code. 

Seriously though, the fact that this is low priority for Confluent makes me sad, because it's not low priority for the people using it (I'm a paying customer, I suggested we use CKP for the support, whoops!) especially when they're pushing hard on Flink which has a strong Python component.

<insert shame nun here>

**Comment by Kyle-sandeman-mrdfood** [comment]
> Especially as type hints can be provided via Typeshed, they don't need to be in the code.

Type hints can also be defined completely separately from runtime code in `.pyi` (stub/interface/header) files. Typeshed is a community effort for providing types to projects that don't do it

**Comment by LiamClarkeNZ** [comment]
I was trying to give an example of how separate they could be if you really
needed them to be, not imply it's the only way to provide them :)

**Comment by ykun92** [comment]
Seems someone already made this
https://github.com/benbenbang/types-confluent-kafka

**Comment by odimko** [comment]
any updates on this? maybe @mhowlett

**Comment by odimko** [comment]
hey there! just checking in if there's any activity on this issue? 

cc @mhowlett

**Comment by j2cry** [comment]
Here is [my attempt](https://github.com/j2cry/confluent-kafka-python/tree/annotations) to make annotations, take a look.

I'm convinced that creating stubs separately from runtime code is writing the same documentation twice.
The best way would be writing annotations in parallel with writing the code.

**Comment by MSeal** [comment]
Related to https://github.com/confluentinc/confluent-kafka-python/issues/1310 -- this is now on radar to address

**Comment by MSeal** [comment]
This is solved via https://github.com/confluentinc/confluent-kafka-python/pull/2123 and will be in the next beta release followed by the minor release after that.
